sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -16,156 +16,195 @@
   {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
-
-    NTSTATUS status = BCryptHashData(
-        md5Context->hashHandle,
-        reinterpret_cast<PBYTE>(const_cast<uint8_t*>(data)),
-        static_cast<ULONG>(length),
-        0);
-    if (!BCRYPT_SUCCESS(status))
+    if (!NT_SUCCESS(
+            m_status = BCryptHashData(
+                m_hashHandle,
+                reinterpret_cast<PBYTE>(const_cast<uint8_t*>(data)),
+                static_cast<ULONG>(length),
+                0)))
     {
-      throw std::runtime_error("BCryptHashData failed");
+      throw std::runtime_error("BCryptHashData failed with code" + m_status);
     }
   }
 
-  std::vector<uint8_t> Md5Hash::OnFinal(const uint8_t* data, std::size_t length)
+  std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length)
   {
     OnAppend(data, length);
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
+
     std::vector<uint8_t> hash;
-    hash.resize(md5Context->hashLength);
-    NTSTATUS status = BCryptFinishHash(
-        md5Context->hashHandle,
-        reinterpret_cast<PUCHAR>(&hash[0]),
-        static_cast<ULONG>(hash.size()),
-        0);
-    if (!BCRYPT_SUCCESS(status))
+    hash.resize(m_hashLength);
+    if (!NT_SUCCESS(
+            m_status = BCryptFinishHash(
+                m_hashHandle,
+                reinterpret_cast<PUCHAR>(&hash[0]),
+                static_cast<ULONG>(hash.size()),
+                0)))
     {
-      throw std::runtime_error("BCryptFinishHash failed");
+      throw std::runtime_error("BCryptFinishHash failed with code" + m_status);
     }
     return hash;
   }
 
-#elif defined(AZ_PLATFORM_POSIX)
-
-  Md5Hash::Md5Hash()
+public:
+  Md5BCrypt()
   {
-    MD5_CTX* md5Context = new MD5_CTX;
-    m_md5Context = md5Context;
-    MD5_Init(md5Context);
+    m_buffer.resize(GetMD5AlgorithmProvider().ContextSize);
+    m_hashLength = GetMD5AlgorithmProvider().HashLength;
+
+    if (!NT_SUCCESS(
+            m_status = BCryptCreateHash(
+                GetMD5AlgorithmProvider().Handle,
+                &m_hashHandle,
+                reinterpret_cast<PUCHAR>(&m_buffer[0]),
+                static_cast<ULONG>(m_buffer.size()),
+                nullptr,
+                0,
+                0)))
+    {
+      throw std::runtime_error("BCryptCreateHash failed with code" + m_status);
+    }
   }
 
-  Md5Hash::~Md5Hash()
+  ~Md5BCrypt()
   {
-    MD5_CTX* md5Context = static_cast<MD5_CTX*>(m_md5Context);
-    delete md5Context;
+    if (m_hashHandle)
+    {
+      BCryptDestroyHash(m_hashHandle);
+    }
   }
+};
 
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
+} // namespace
+Azure::Core::Cryptography::Md5Hash::Md5Hash() : m_implementation(std::make_unique<Md5BCrypt>()) {}
+
+#elif defined(AZ_PLATFORM_POSIX)
+
+class Md5OpenSSL : public Azure::Core::Cryptography::Hash {
+private:
+  std::unique_ptr<MD5_CTX> m_context;
```

Comment:
Do we need a dtor to delete this `m_context`, or will the unique ptr help us here and clean up as soon as `Md5OpenSSL` is out of scope?

Reply:
> will the unique ptr help us here and clean up

Yes